### PR TITLE
Optimizations

### PR DESCRIPTION
**Pull request by Shawdooow**
Use unsafe magic to cast memory for free!

Also added new `SubArray` function

### CODE REVIEW

**Comment by Shawdooow** [comment]
Actually to my extreme amazement testing in `.net6.0` has revealed that the crazy `if` statements were as fast as an `Unsafe.As` cast and MemoryMarshal is slower.

Could be different for the version of net this project is compiled against, definitely warrants further investigation...

image

**Comment by Shawdooow** [comment]
Using `Unsafe.As` wasn't any faster than all those crazy if statements because the JIT was working overtime on it. The if statements are slower when the JIT can't guarantee that an array with 3 or greater items is being fed into it.
*Note the bottom function shown is using floats NOT bytes*
image

**Comment by Shawdooow** [comment]
I can't actually compile this repo to test how these changes interact with each other but I can say that I tested them individually and they all appear to work.

File: vke/src/Utils.cs [hunk 1]
```diff
@@ -67,88 +68,61 @@ public static XmlAttributeOverrides GetXmlOverrides ()
...
+		public static void FromFloatArray (ref Quaternion q, float[] floats) {
+			q = Unsafe.As<float[], Quaternion[]>(ref floats)[0];
 		}
 		/// <summary>
 		/// Populate a Vector2 with values from a byte array starting at offset
 		/// </summary>
 		public static void FromByteArray (ref Vector2 v, byte[] byteArray, int offset) {
-            v.X = BitConverter.ToSingle (byteArray, offset);
-            v.Y = BitConverter.ToSingle (byteArray, offset + 4);
+			v = Unsafe.As<byte[], Vector2[]>(ref Unsafe.AsRef(byteArray.SubArray(offset, 8)))[0];
```
Comment (Shawdooow) [inline vke/src/Utils.cs:89]: I don't know how large the byte arrays that typically get passed into this are but if they are typically large (>64 bits) I would consider passing them in as a `ref`

**Comment by jpbruyere** [comment]
Why can't you compile the repo? I can help surely.

Can you guaranty that Vector structs and Quaternion Lenght (sizeof) is always correct. (see https://stackoverflow.com/questions/42695067/unsafe-as-from-byte-array-to-ulong-array).

I have already implemented modifications in other project of mine with Span and ReadOnlySpan, and I've not yet optimized vke. I knew it was to be done, thanks for the time you spend on this.

**Review by jpbruyere — APPROVED** [review]
build and testing ok, Marshal.SizeOf seems ok.

**Comment by Shawdooow** [comment]
The size and layout of `Vector`s and `Quaternion`s in memory shouldn't ever change.

It should also be noted that despite `W` at the top of those structs in code in memory they are last (which I only learned doing this pr)

**Comment by Shawdooow** [comment]
On a semi-unrelated note, do you have a slack or discord server related to your vulkan repos?

**Comment by jpbruyere** [comment]
I have just invited you into my Gitter channel (https://gitter.im/CSharpRapidOpenWidgets/) (I should rename it)

**Pull request merged**